JavaScript/JavaScriptSDK.Tests/CheckinTests/Context/Session.tests.ts

```diff
@@ -120,6 +128,88 @@ class SessionContextTests extends TestClass {
+                if (window.localStorage) {
+                    // setup
+                    var actualCookieName: string;
+                    var actualCookieValue: string;
+                    var newGuidStub = sinon.stub(Microsoft.ApplicationInsights.Util, "newGuid",() => "newGuid");
+                    var getCookieStub = sinon.stub(Microsoft.ApplicationInsights.Util, "getCookie",() => "");
+                    var setCookieStub = sinon.stub(Microsoft.ApplicationInsights.Util, "setCookie",(cookieName, cookieValue) => { });
+
+                    // act
+                    var sessionManager = new Microsoft.ApplicationInsights.Context._SessionManager(null,() => { });
+                    sessionManager.update();
+                    sessionManager.backup();
+
+                    // verify
+                    Assert.ok(localStorage["ai_session"], "ai_session storage is set");
+
+                    Assert.equal(3, localStorage["ai_session"].split('|').length, "Cookie value before expiration should include user id, acq date and renew date");
+                    Assert.equal("newGuid", localStorage["ai_session"].split('|')[0], "First part of cookie value should be new user id guid");
+
+                    // cleanup
+                    getCookieStub.restore();
+                    setCookieStub.restore();
+                    newGuidStub.restore();
+                } else {
+                    // this might happen on IE when using a file:// url
+                    Assert.ok(true, "browser does not support local storage in current environment");
+                }
+            }
+        });
+
+        this.testCase({
+            name: "SessionContext: session manager can recover old session id and isFirst state from lost cookies when localStorage is available",
+            test: () => {
+                // A more sophisticated approach of emulating document.cookie is needed to support multiple cookies
+                Microsoft.ApplicationInsights.Util["document"] = this.generateEmulatedCookieStore();
+
+                // Initialize our user and session cookies, as well as local storage
+                new Microsoft.ApplicationInsights.Context.User(undefined);
+                var sessionManager = new Microsoft.ApplicationInsights.Context._SessionManager(null,() => { });
+                sessionManager.update();
+                sessionManager.backup();
+                var sessionId = sessionManager.automaticSession.id;
+
+                // Lose the session cookie but not the user cookie
+                Microsoft.ApplicationInsights.Util["document"].cookie = "ai_session=; expires=Thu, 01 Jan 1970 00:00:01 GMT;";
+                new Microsoft.ApplicationInsights.Context.User(undefined);
+                var sessionManager = new Microsoft.ApplicationInsights.Context._SessionManager(null,() => { });
+                sessionManager.update();
+                sessionManager.backup();
+
+                if (window.localStorage) {
```

Comment:
This if is actually so that browsers that do not support local storage, like IE when using file:// urls, do not cause improper test failures